### PR TITLE
adds a middleware to add the backstage identity

### PR DESCRIPTION
**Pull request by punkle**
Currently the scaffolder backend is parsing the backend auth token as a JWT token. This works well where:

- the different parts of the backstage backend are hosted on the same server / network trust domain
- the same server that is generating the tokens is also verifying them

However where the scaffolder and catalog services are hosted separately, then the token produced by backstage must be understood by the infrastructure forwarding the requests between the components of backstage. This would require a centrally managed certificate trust infrastructure or a central token issuer or something else. This is a large architectural ask.

This change expands the flexibility of the backstage identity platform by allowing any format of token for building the backstage identity for the plugins to read.

This patch also causes the scaffolder backend to use the identity from the new middleware.

It is configured by adding `packages/backend/src/plugins/authentication-middleware.ts`:

```typescript
import { RequestHandler } from 'express';
import { PluginEnvironment } from '../types';
import {
  createMiddleware,
  jwtMiddlewareProvider,
} from '@backstage/plugin-authentication-middleware-backend';

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<RequestHandler> {
  return await createMiddleware({
    logger: env.logger,
    authenticationMiddlewareProvider: jwtMiddlewareProvider,
  });
}
```

In `packages/backend/index.ts`, somewhere close to the top of the stack.

```typescript
  ...
  const authenticationMiddlewareEnv = useHotMemoize(module, () =>
    createEnv('authentication-middleware'),
  );
  ...
  const apiRouter = Router();
  apiRouter.use(await authenticationMiddleware(authenticationMiddlewareEnv));
```

In the backend plugins you can access the identity as follows:

```typescript
  router.get('/something', async (req: AuthenticatedBackstageRequest, res) => {
    const identity = req.backstage?.identity;
    res.send(identity);
  });
```

Signed-off-by: Brian Fletcher <brian@roadie.io>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AuthenticationMiddlewareProvider } from '.';
+import { JsonValue } from '@backstage/types';
+import { InputError, stringifyError } from '@backstage/errors';
+
+export const unverifiedJWTMiddlewareProvider: AuthenticationMiddlewareProvider =
```
Comment (punkle) [inline plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts:21]: This is what the scaffolder is currently doing. Extracting the jwt token from the header, parsing it and extracting the ent and sub payload fields and setting them as the identity.

**Comment by punkle** [comment]
codeql whack a mole.

File: plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  async req => {
+    if (!req.headers?.authorization) {
+      return undefined;
+    }
+
+    try {
+      if (!req.headers.authorization.startsWith('Bearer ')) {
+        return undefined;
+      }
+      const token = req.headers.authorization.replace(/^Bearer /, '');
```
Comment (kuangp) [inline plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts:37]: Could this utilize [`getBearerTokenFromAuthorizationHeader`](https://github.com/backstage/backstage/blob/master/plugins/auth-node/src/getBearerTokenFromAuthorizationHeader.ts)?

File: plugins/scaffolder-backend/src/service/router.ts [hunk 3]
```diff
@@ -180,14 +183,17 @@ export async function createRouter(
...
       });
-      const { token, entityRef: userEntityRef } = parseBearerToken(
-        req.headers.authorization,
-      );
+      const token =
+        req.backstage?.identity?.token ||
+        parseBearerToken(req.headers.authorization).token;
+      const userEntityRef =
+        req.backstage?.identity?.identity?.userEntityRef ||
+        parseBearerToken(req.headers.authorization).entityRef;
```
Comment (kuangp) [inline plugins/scaffolder-backend/src/service/router.ts:196]: Maybe out of scope of this change specifically but my understanding was that the new pattern for obtaining/verifying the (default jwt) user token is:
```ts
const token = getBearerTokenFromAuthorizationHeader(req.header('authorization'));
  
const user = await identity.authenticate(token);
```

**Review by kuangp — COMMENTED** [review]
hmm .. if the token format becomes configurable, would that raise concerns that all backend plugins needs to be able to understand the given token ? 

eg. the auth middleware can be configured to correctly parse the token and attach a `req.backstage.identity` but all backend plugins would either need to be updated to be aware of and utilize this new prop or know how to parse the token themselves 

Wouldn't configuration of the `auth-backend` plugin and `IdentityClient` (assuming plugins are using this and not parsing the raw token themselves) itself allow for customization of the desired token format ? or does that still run into the centralized token issuer issue? 

Let me know if I've misunderstood design here 😅

File: plugins/scaffolder-backend/src/service/router.ts (hunk 3 above)
Comment (punkle) [reply plugins/scaffolder-backend/src/service/router.ts:196]: That looks like a better pattern. Let me implement a provider that does this. I would be inclinded to delete the parseBearerToken function in the scaffolder but I would be afraid to break anyone. I left it here to give people time to uptake the middleware.

File: plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts (hunk 2 above)
Comment (punkle) [reply plugins/authentication-middleware/src/providers/unverifiedJWTMiddlewareProvider.ts:37]: Probably yes, I will look into this.

**Comment by punkle** [comment]
> hmm .. if the token format becomes configurable, would that raise concerns that all backend plugins needs to be able to understand the given token ?
> 
> eg. the auth middleware can be configured to correctly parse the token and attach a `req.backstage.identity` but all backend plugins would either need to be updated to be aware of and utilize this new prop or know how to parse the token themselves
> 
> Wouldn't configuration of the `auth-backend` plugin and `IdentityClient` (assuming plugins are using this and not parsing the raw token themselves) allow for customization of the desired token format ? or does that still run into the centralized token issuer issue?
> 
> Let me know if I've misunderstood the design here 😅

I want to suggest that all plugins, if they need the user identity of the service/user would look for the identity in the req.backstage.identity field populated by this middleware. This would mean that the token would be parsed once by the middleware and the backend plugins would not need to have this knowledge.

I was working on the assumption that backstage backend plugins, would operate on the backstage identity and not on the the token itself. So I could have got this assumption wrong. Can you explain an example of a case where a backend plugin might need to understand the token format?

`auth-backend` allows issuing any format of token. However today the `scaffolder-backend` is parsing it as a jwt token. I am proposing implementing this middleware, so that there is a pattern to follow to parse the token consistently across the application meaning the scaffolder-backend would not need to parse the token by itself in the first place.

The IdentityClient requires the issuer and the algorithms. I suppose this could be constructed in the backend and passed down to all of the plugins and that could be used to figure out the user. But currently the IdentityClient assumes that the token is jwt also. Maybe it can be changed to an interface that can be provided? Either would work I think.

**Comment by punkle** [comment]
Thanks for your review @kuangp

**Comment by kuangp** [comment]
> I was working on the assumption that backstage backend plugins, would operate on the backstage identity and not on the the token itself. So I could have got this assumption wrong. Can you explain an example of a case where a backend plugin might need to understand the token format?

So what I meant was for other plugins that have not been updated and are still manually parsing the token (like the scaffolder-backend prior to this change), they would not be able to understand a custom token format if said plugin has not been updated yet. 

So similar to your proposal, I wonder if it makes sense as an alternative, all plugins are updated to handle tokens via `const user = await identity.authenticate(token);` to allow the `IdentityClient` be responsible for knowing how to parse out the user. That way, instead of having to install/configure this middleware + updating all plugins to read the new `req.backstage?.identity` field, all plugins utilize the `IdentityClient` (some of which already are) and integrators customize the `IdentityClient` through some interface as you suggested.

The benefits here would be not having to install an additional middleware package and not having each plugin loosely coupled to the semantics of the `req.backstage` prop provided by said middleware.

**Comment by Rugvip** [comment]
Yep having discussed similar things as a part of #11611 I'm leaning towards `const data = <service>.<method>(req)` being much better than decorating the request object directly. Especially with typescript it's essentially just the same amount of "work" too. You can even go and cache the result per request within the service implementation if it's doing something expensive.

**Comment by punkle** [comment]
I closed this in favour of the better suggestion.

**Pull request closed without merging**